src/Processors/Streaming/ResizeProcessor.cpp

```diff
@@ -0,0 +1,456 @@
+}
+
+namespace Streaming
+{
+ShrinkResizeProcessor::ShrinkResizeProcessor(const Block & header, size_t num_inputs)
+    : IProcessor(InputPorts(num_inputs, header), OutputPorts(1, header))
+    , log(&Poco::Logger::get("ShrinkResizeProcessor"))
+{
+    assert(num_inputs > 0);
+}
+
+IProcessor::Status ShrinkResizeProcessor::prepare(const PortNumbers & updated_inputs, const PortNumbers & /*updated_outputs*/)
+{
+    if (unlikely(!initialized))
+    {
+        initialized = true;
+
+        for (auto & input : inputs)
+        {
+            assert(input.getOutputPort().getProcessor().isStreaming());
+            input.setNeeded();
+            input_ports.push_back({.port = &input, .status = InputStatus::NeedData, .watermark = INVALID_WATERMARK});
+        }
+    }
+
+    /// Update inputs
+    for (const auto & input_number : updated_inputs)
+    {
+        auto & input = input_ports[input_number];
+        if (input.port->isFinished())
+        {
+            if (input.status != InputStatus::Finished)
+            {
+                input.status = InputStatus::Finished;
+                ++num_finished_inputs;
+            }
+            continue;
+        }
+
+        if (input.port->hasData())
+        {
+            if (input.status != InputStatus::HasData)
+            {
+                input.status = InputStatus::HasData;
+                inputs_with_data.push(input_number);
+            }
+        }
+    }
+
+    if (num_finished_inputs == inputs.size())
+    {
+        for (auto output : outputs)
+            output.finish();
+
+        return Status::Finished;
+    }
+
+    /// Check output can push
+    auto & output = outputs.front();
+    if (output.isFinished())
+    {
+        for (auto & input : inputs)
+            input.close();
+
+        return Status::Finished;
+    }
+
+    if (!output.canPush())
+        return Status::PortFull;
+
+    /// Check inputs has data
+    for (auto & input_port : input_ports)
+    {
+        if (input_port.status == InputStatus::NeedData)
+            input_port.port->setNeeded();
+    }
+
+    if (!inputs_with_data.empty())
+    {
+        auto & input_with_data = input_ports[inputs_with_data.front()];
+        inputs_with_data.pop();
+
+        auto data = input_with_data.port->pullData(true);
+        if (updateAndAlignWatermark(input_with_data, data.chunk))
+        {
+            /// Do nothing
+        }
+        else
+            input_with_data.status = InputStatus::NeedData;
+
+        output.pushData(std::move(data));
+        return Status::PortFull;
+    }
+
+    return Status::NeedData;
+}
+
+bool ShrinkResizeProcessor::updateAndAlignWatermark(InputPortWithStatus & input_with_data, Chunk & chunk)
+{
+    if (!chunk.hasWatermark())
+        return false;
+
+    bool updated = false;
+    auto new_watermark = chunk.getWatermark();
+    if (new_watermark > input_with_data.watermark || (input_with_data.watermark == TIMEOUT_WATERMARK && new_watermark >= aligned_watermark))
```

Comment:
How the `new_watermark > input_with_data.watermark` can be false?

Reply:
When we do alignment, watermarks can be from different data sources, so it is possible new_watermark can be from a fast input.